docs/developer/architecture/authentication.rst

```diff
@@ -0,0 +1,97 @@
+Authentication Architecture Document
+====================================
+
+Context:
+--------
+Currently, when we deploy MetalK8s we pre provision super admin user with admin
+password. Meaning anyone wanting to use K8S/Salt APIs should authenticate using
+this single super admin user. Another way to access the APIs is by using K8S
+admin certificate that is stored in /etc/kubernetes/admin.conf. We could also
+manually provisioned other users and credentials as well as roles binding
+but the way the functionality is accessible currently is too complex and not
+enough secured since user / password are stored in clear text.
+This is not good enough in term of security and flexibility.
+
+We would like at least to be able to add different users with different
+credentials and ideally to integrate K8S authentication system with external
+identity provider.
+
+Managing K8S role binding between user / groups High level roles and K8S roles
+is not part of this specification.
+
+Requirements:
+-------------
+
+Basically we are talking about:
+
+- Being able to provision users with an embedded Identity provider (IDP)
+- Being able to integrate with an external IDP
+
+`Dex <https://github.com/dexidp/dex>`_ seems to be the preferred solution for
+now, for both local identity provider or proxy to external IDP.
+Integration with LDAP or Microsoft AD are the most important ones to support.
+
+User Stories:
+-------------
+**User Management with Dex as an IDP**
+
+As an IT Generalist, I want to provision/edit usersa and high level roles.
+group provisioning not supported by Dex from what I understand.
+The Product high level roles are:
+
+- Cluster Admin role
+- Solution Admin role
+- Read Only
+
+This is done from CLI with well documented procedure.
+Entered passwords are never visible and encrypted when stored in DEX DB.
+The CLI tool enables to add / delete and edit passwords and roles.
+
+The CLI tool is used right after bootstrap node deployment to provision the
+very first user (which won't be provisioned anymore by default)
```

Comment:
I would argue that this will not be the first user to be provisioned.

I still think that the `Scality-Dex` superuser needs to be statically defined else we will have no way to connect Dex to the API-server.
Adding to this, the user experience becomes even more awful when we have to ask users to configure the API-server/Dex integration and force reboot the API-server.
https://github.com/dexidp/dex/blob/master/Documentation/kubernetes.md#configuring-the-openid-connect-plugin

Reply:
No, you don't need a user to have K8s API accept OIDC tokens issued from Dex. As explained in your docs, the setup requires certificates to be generated (that's going to be done with Salt and the `ca:minion`), said certs must then be exposed to API Server, and that should be enough.

Reply:
Dex connecting to K8s API is using a `ServiceAccount`, so no need to provision any OIDC user. And as @gdemonet said, you can set up K8s API to accept OIDC tokens *even* when the OIDC service is not (yet) running. No need for any restart.

Which is a good thing, otherwise your whole cluster would stop operating when there's an issue with some external component, i.e. the OIDC IdP in this case?

To define new users (in the case of Dex), we could either use its gRPC API (though I have some concerns about authn towards said API...), or work at the `ConfigMap` level, which is not an issue since we *have* `/etc/kubernetes/admin.conf` available when the user management script is executed. No need for any OIDC 'user' either.